### DISCUSSION TITLE
ch6/ch6-02

### DISCUSSION THREAD

**Discussion by giscus** [gopl-zh/gopl-zh.github.com#172] [General]
# ch6/ch6-02

<The Go Programming Language>中文版

https://golang-china.github.io/gopl-zh/ch6/ch6-02.html

**Comment by FanXieGangN** [comment]
最后一段原文：
Because url.Values is a map type and a map refers to its key/value pairs indirectly, any
updates and deletions that url.Values.Add makes to the map elements are visible to the
caller. However, as with ordinary functions, any changes a method makes to the reference
itself, like setting it to nil or making it refer to a different map data structure, will not be
reflected in the caller.

**Comment by KrxkGit** [comment]
我按自己理解做个总结吧：（姑且用对象来做说明吧）
1. 按值传递，调用函数时会复制该对象与传递函数形参，在函数内部修改该对象需要通过return来返回（否则将只相当于函数内部的临时变量）
2. 按指针传递，调用函数时将传递对象指针，然后传递函数形参，在函数内部的修改就是对指针指向的内存的修改。

测试代码如下：
import "fmt"

type Number struct {
	sum int
}

func (n Number) Add(i int) Number {
	n.sum += i
	//res = n
	return n
}

func (np *Number) AddbyPtr(i int) {
	np.sum += i
}

func Test() {
	a := Number{1}
	b := Number{1}

	a.Add(1)
	fmt.Printf("a: %d\n", a.sum)
	b.AddbyPtr(1)
	fmt.Printf("b: %d\n", b.sum)

	fmt.Println()

	c := a
	c.Add(1)
	fmt.Printf("a: %d\n", a.sum)
	fmt.Printf("c: %d\n", c.sum)

	e := a.Add(1)
	fmt.Printf("e: %d\n", e.sum)

	fmt.Println()

	d := b
	d.AddbyPtr(1)
	fmt.Printf("b: %d\n", b.sum)
	fmt.Printf("d: %d\n", d.sum)
}

运行结果：
a: 1
b: 2
    
a: 1
c: 1
e: 2
    
b: 2
d: 3

**Reply by KrxkGit** [reply]
同时可以注意到最后一组，d := b 会发生结构体的拷贝，但这不是发生在AddbyPtr调用的时候的，增加一组测试代码：
fmt.Println()
f := &b
f.AddbyPtr(1)
fmt.Printf("b: %d\n", b.sum)
fmt.Printf("f: %d\n", f.sum)

运行结果如下：
b: 3
f: 3

**Comment by xm-xw** [comment]
**这写的太乱了，还不如直接去搜：golang的值接收器与指针接收器**
### 1、本质就是传值与传地址的区别
    当然，go语言中只有传值，因为传地址本质上也是将地址的副本传给函数，函数再通过地址来对地址上的值进行修改
### 2、体现在这里主要区别就是：
    1.如果是指针接收器，通过调用方法会将接收器的【地址的拷贝副本】传给函数（方法），可以对接收器进行寻址，进而修改接收器的值
    2.如果是值接收器，通过调用方法会将接收器的【值的拷贝副本】传给函数（方法），改变的也只是方法内部的接收器的拷贝副本罢了
### 3、那么就只需要考虑使用场景和使用注意点了
#### 1、使用场景
    1.如果需要修改接收器的值就使用指针接收器，如果不需要就使用值接收器
    2.一般来说我们建议使用指针接收器，一方面是传递地址【省内存】，还有一方面得益于go语言对内存安全的机制，不需要手动释放，可以放心使用指针
#### 2、注意点
    1.主要针对一下用法，要求ScaleBy的接收器不能是指针接收器类型（*Point），因为这样的Point变量是一个临时变量，生存期只有这一行，ScaleBy内部无法对其进行寻址
```go
Point{1, 2}.ScaleBy(2)
```

**Comment by lunarlinger** [comment]
差别就是类似于cpp中的“值传递”和“引用传递”的差别吧
